lib/heap/cmpctmalloc/cmpctmalloc.c

```diff
@@ -0,0 +1,723 @@
+	// GCC intrinsic count-leading-zeros.
+	// Eg. 128-255 has 24 leading zeros and we want row to be 4.
+	unsigned row = sizeof(size_t) * 8 - 4 - __builtin_clzl(size);
+	// For row 4 we want to shift down 4 bits.
+	unsigned column = (size >> row) & 7;
+	int row_column = (row << 3) | column;
+	row_column += increment;
+	size = (8 + (row_column & 7)) << (row_column >> 3);
+	*rounded_up_out = size;
+	// We start with 14 buckets, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96,
+	// 104, 112, 120.  Then we have row 4, sizes 128 and up, with the
+	// row-column 8 and up.
+	int answer = row_column + 14 - 32;
+	DEBUG_ASSERT(answer < NUMBER_OF_BUCKETS);
+	return answer;
+}
+
+// Round up size to next bucket when allocating.
+static int size_to_index_allocating(size_t size, size_t *rounded_up_out)
+{
+	size_t rounded = ROUNDUP(size, 8);
+	return size_to_index_helper(rounded, rounded_up_out, -8, 1);
+}
+
+// Round down size to next bucket when freeing.
+static int size_to_index_freeing(size_t size)
+{
+	size_t dummy;
+	return size_to_index_helper(size, &dummy, 0, 0);
+}
+
+inline header_t *tag_as_free(void *left)
+{
+	return (header_t *)((uintptr_t)left | 1);
+}
+
+inline bool is_tagged_as_free(header_t *header)
+{
+	return ((uintptr_t)(header->left) & 1) != 0;
+}
+
+inline header_t *untag(void *left)
+{
+	return (header_t *)((uintptr_t)left & ~1);
+}
+
+inline header_t *right_header(header_t *header)
+{
+	return (header_t *)((char *)header + header->size);
+}
+
+inline static void set_free_list_bit(int index)
+{
+	theheap.free_list_bits[index >> 5] |= (1u << (31 - (index & 0x1f)));
+}
+
+inline static void clear_free_list_bit(int index)
+{
+	theheap.free_list_bits[index >> 5] &= ~(1u << (31 - (index & 0x1f)));
+}
+
+static int find_nonempty_bucket(int index)
+{
+	uint32_t mask = (1u << (31 - (index & 0x1f))) - 1;
+	mask = mask * 2 + 1;
+	mask &= theheap.free_list_bits[index >> 5];
+	if (mask != 0) return (index & ~0x1f) + __builtin_clz(mask);
+	for (index = ROUNDUP(index + 1, 32); index <= NUMBER_OF_BUCKETS; index += 32) {
+		mask = theheap.free_list_bits[index >> 5];
+		if (mask != 0u) return index + __builtin_clz(mask);
+	}
+	return -1;
+}
+
+bool is_start_of_os_allocation(header_t *header)
+{
+	uintptr_t address = (uintptr_t)header;
+	if ((address & (PAGE_SIZE - 1)) != 0) return false;
+	if (header->size != sizeof(header_t)) return false;
+	return header->left == untag(NULL);
+}
+
+static void create_free_area(void *address, void *left, size_t size, free_t **bucket)
+{
+	free_t *free_area = (free_t *)address;
+	free_area->header.size = size;
+	free_area->header.left = tag_as_free(left);
+	if (bucket == NULL) {
+		int index = size_to_index_freeing(size);
+		set_free_list_bit(index);
+		bucket = &theheap.free_lists[index];
+	}
+	free_t *old_head = *bucket;
+	if (old_head != NULL) old_head->prev = free_area;
+	free_area->next = old_head;
+	free_area->prev = NULL;
+	*bucket = free_area;
+	theheap.remaining += size;
+#ifdef DEBUG
```

Comment:
Might be nice to move this to some sort of local #define, like HEAP_DEBUG then can be set at the top of the file.